src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.DocumentChangeTracker.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    internal partial class RequestExecutionQueue
+    {
+        /// <summary>
+        /// Keeps track of changes to documents that are opened in the LSP client. Calls MUST not overlap, so this
+        /// should be called from a mutating request handler. See <see cref="RequestExecutionQueue"/> for more details.
+        /// </summary>
+        internal class DocumentChangeTracker : IWorkspaceService
+        {
+            private readonly Dictionary<(Workspace, DocumentId), Document> _trackedDocuments = new();
+
+            internal async Task StartTrackingAsync(Document document, string contents, CancellationToken cancellationToken)
```

Comment:
I feel a little weird about how didOpen is handled vs didChange in the tracker.  In the didChange handler, we first update the document text inside the handler, then tell the DocumentChangeTracker about the updated text.  Whereas in didOpen, we're doing all the document operations inside the DocumentChangeTracker.

Perhaps instead we pass in the new document text here?  Then it becomes more similar to didChange, we first do the document operations (create new document from text) inside the handler, then update the DocumentChangeTracker

Reply:
Yes, thank you! This makes so much sense and makes everything nicer, I was too familiar with the code to see it. Love it!